kernel/comps/virtio/src/device/input/device.rs

```diff
@@ -181,44 +185,133 @@ impl InputDevice {
-        loop {
-            let Some(event) = self.pop_pending_event() else {
-                return Some(());
+    fn handle_irq(&self) {
+        // Returns ture if there may be more events to handle
+        let handle_event = |event: &EventBuf| -> bool {
+            let event: VirtioInputEvent = {
+                let mut reader = event.reader();
+                reader.read_val()
             };
+
             match event.event_type {
-                0 => return Some(()),
+                0 => return false,
                 // Keyboard
                 1 => {}
                 // TODO: Support mouse device.
-                _ => continue,
+                _ => return true,
             }
+
             let status = match event.value {
                 1 => KeyStatus::Pressed,
                 0 => KeyStatus::Released,
-                _ => return Some(()),
+                _ => return false,
             };
+
             let event = InputEvent::KeyBoard(Key::try_from(event.code).unwrap(), status);
             info!("Input Event:{:?}", event);
 
             let callbacks = self.callbacks.lock();
             for callback in callbacks.iter() {
-                callback.call((event,));
+                callback(event);
             }
+
+            true
+        };
+
+        // one interrupt may contains serval input, so it should loop
+        while self.pop_pending_event(&handle_event) {}
+    }
+
+    /// Negotiate features for the device specified bits 0~23
+    pub(crate) fn negotiate_features(features: u64) -> u64 {
+        assert_eq!(features, 0);
+        0
+    }
+}
+
+/// A event table consists of many event buffers,
+/// each of which is large enough to contain a `VirtioInputEvent`.
+#[derive(Debug)]
+struct EventTable {
+    stream: DmaStream,
+    num_events: usize,
+}
+
+impl EventTable {
+    fn new(num_events: usize) -> Self {
+        assert!(num_events * mem::size_of::<VirtioInputEvent>() <= PAGE_SIZE);
+
+        let vm_segment = {
+            let mut options = VmAllocOptions::new(1);
+            options.is_contiguous(true);
+            options.alloc_contiguous().unwrap()
+        };
+
+        let default_event = VirtioInputEvent::default();
+        for idx in 0..num_events {
+            let offset = idx * EVENT_SIZE;
+            vm_segment.write_val(offset, &default_event).unwrap();
+        }
+
+        let stream = DmaStream::map(vm_segment, DmaDirection::FromDevice, false).unwrap();
+        Self { stream, num_events }
+    }
+
+    fn get(&self, idx: usize) -> EventBuf<'_> {
+        assert!(idx < self.num_events);
+
+        let offset = idx * EVENT_SIZE;
+        EventBuf {
+            event_table: self,
+            offset,
+            size: EVENT_SIZE,
         }
     }
 
+    const fn num_events(&self) -> usize {
+        self.num_events
+    }
+}
+
+const EVENT_SIZE: usize = core::mem::size_of::<VirtioInputEvent>();
+
+/// A buffer stores exact one `VirtioInputEvent`
+struct EventBuf<'a> {
```

Comment:
This `EventBuf` is every much like `SafePtr`. I am thinking if `SafePtr` can be extended to satisfy this need or a a new `DmaPtr` should be added.

Reply:
I have extended `SafePtr` to support `DmaStream`. Then EventBuf can be replaced with `SafePtr<VirtioInputEvent, DmaStream>`

```rust
impl<T, M: HasDaddr, R> HasDaddr for SafePtr<T, M, R> {
    fn daddr(&self) -> Daddr {
        self.offset + self.vm_obj.daddr()
    }
}

impl<T, R> SafePtr<T, DmaStream, R> {
    /// Synchronize the object in the streaming DMA mapping
    pub fn sync(&self) -> Result<()> {
        self.vm_obj
            .sync(self.offset..self.offset + core::mem::size_of::<T>())
    }
}
```